cloudformation/xtdb_msk.yml

```diff
@@ -0,0 +1,161 @@
+AWSTemplateFormatVersion: '2010-09-09'
+
+Parameters:
+  VpcId: 
+    Description: ID of the VPC to host the MSK cluster on
+    Type: AWS::EC2::VPC::Id
+
+  SecurityGroupId:
+    Description: Group ID of the security group to be used by the ECS nodes
+    Type: AWS::EC2::SecurityGroup::Id
+
+  MSKClusterName:
+    Type: String
+    Default: xtdb-cluster-kafka
+    Description: Enter the desired name of the MSK kafka cluster used as the XTDB Kafka Log
+
+  MSKVolumeSize:
+    Type: Number
+    Default: 100
+    Description: The size in GiB of the EBS volume for the data drive on each broker node of the kafka cluster.
+
+Resources:
+  PrivateSubnetMSKOne:
+    Type: AWS::EC2::Subnet
+    Properties:
+      AvailabilityZone:
+         Fn::Select:
+         - 0
+         - Fn::GetAZs: {Ref: 'AWS::Region'}
+      VpcId: !Ref VpcId
+      CidrBlock: 10.192.20.0/24
+      MapPublicIpOnLaunch: false
+
+  PrivateSubnetMSKTwo:
+    Type: AWS::EC2::Subnet
+    Properties:
+      AvailabilityZone:
+         Fn::Select:
+         - 1
+         - Fn::GetAZs: {Ref: 'AWS::Region'}
+      VpcId: !Ref VpcId
+      CidrBlock: 10.192.21.0/24
+      MapPublicIpOnLaunch: false
+
+  PrivateRouteTable:
+    Type: AWS::EC2::RouteTable
+    Properties:
+      VpcId: !Ref VpcId
+  
+  PrivateSubnetMSKOneRouteTableAssociation:
+    Type: AWS::EC2::SubnetRouteTableAssociation
+    Properties:
+      RouteTableId: !Ref PrivateRouteTable
+      SubnetId: !Ref PrivateSubnetMSKOne
+
+  PrivateSubnetMSKTwoRouteTableAssociation:
+    Type: AWS::EC2::SubnetRouteTableAssociation
+    Properties:
+      RouteTableId: !Ref PrivateRouteTable
+      SubnetId: !Ref PrivateSubnetMSKTwo
+  
+  MSKSecurityGroup:
+    Type: AWS::EC2::SecurityGroup
+    Properties:
+      GroupDescription: MSK Security Group
+      VpcId: !Ref VpcId
+      SecurityGroupIngress:
+      - IpProtocol: tcp
+        FromPort: 2181
+        ToPort: 2181
+        SourceSecurityGroupId: !Ref SecurityGroupId
+      - IpProtocol: tcp
+        FromPort: 9098
+        ToPort: 9098
+        SourceSecurityGroupId: !Ref SecurityGroupId
+      - IpProtocol: tcp
+        FromPort: 9094
+        ToPort: 9094
+        SourceSecurityGroupId: !Ref SecurityGroupId
+      - IpProtocol: tcp
+        FromPort: 9092
+        ToPort: 9092
+        SourceSecurityGroupId: !Ref SecurityGroupId
+  
+  MSKSelfIngressAllowRule:
```

Comment:
does this need a separate resource, or can it be included under `MSKSecurityGroup` with the others?

Reply:
I think since it's self referential (ie, security group pointing to itself) it _has_ to be done separately: https://stackoverflow.com/questions/46962263/cloudformation-securitygroup-circular-reference